### ISSUE TITLE
Original formatting does not hold after removing the text(hit backspace)

### ISSUE THREAD

**Issue by geronimol** [singerdmx/flutter-quill#2110]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/singerdmx/flutter-quill/issues)

### Flutter Quill version

10.1.10

### Steps to reproduce

1. Select a font/font size and pick a color from the toolbar.
2. Write a word/character.
3. Remove the word/character.
4. Start typing again.

![Captura de pantalla 2024-08-14 a la(s) 11 46 10](https://github.com/user-attachments/assets/d06dd6b4-75e2-4bec-a765-5680acf50745)
![Captura de pantalla 2024-08-14 a la(s) 11 46 42](https://github.com/user-attachments/assets/48a18cd4-1d67-4bb2-864e-43e242a23ee0)

### Expected results

The new text should keep the original formatting(styles).

### Actual results

The new text does not have any selected style.

### Code sample

<details><summary>Code sample</summary>

```dart
[Paste your code here]
```

</details>

### Additional Context

<details>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

<details><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

**Comment by AtlasAutocode** [comment]
I understand what you are saying and agree that what you are asking sounds reasonable. But as I started to look at how to do it, I found out that there are many conflicting problems.

Attributes are associated with text. When you remove the text, you remove the attributes.
QuillController has toggledStyle which could serve as a sort-of cache for selected style.

The difficulty is deciding the rules for when toggledStyle **_should not be_** updated to solve your issue, and when it **_should be_** updated to reflect what the user expects to happen.

In your case, you remove the styled text. No text means no attributes so toggledStyle should not be updated and should keep the styles last set. So, when should toggledStyle be updated? How do we recognize when it should not be updated?

It is easy to solve this issue when the text is at the start of the document. It is more difficult to decide what should happen if the text is somewhere else in the document and is surrounded by other text that is styled differently. Do you show the style for what you just deleted (not visible to the user) or show the style for what the user sees on the document display?

The relevant part of the code is QuillController _updateSelection. This updates toggledStyle whenever the selection changes (which is always!).

Can you take a look at the method and see if you can come up with some rules that we can apply to solve this issue?

**Comment by geronimol** [comment]
@AtlasAutocode thanks for the information. I've been looking at that function and testing, but unfortunately I don't have knowledge on how I can achieve that without messing up other behaviors in that section, and there are many fields that I would need to learn how they work.

I think this should be considered a bug, because this is how most text editors work, and setting all your styles each time you remove the text isn't good. Also this is a problem when you have a default style set, right?

**Comment by AtlasAutocode** [comment]
> there are many fields that I would need to learn how they work

Welcome to the club. The code is very complex and not always intuitive.
> how I can achieve that without messing up other behaviors

That is the problem. Which is why we need a consistent rule on how style is updated when the selection changes.

Playing around in Microsoft Word (as an example of a widely used word processor), you can enter styled text in the middle of plain text. You can then remove styled letters by backspacing. When the last styled character is removed, the styling IS retained. However, if you move the cursor to the start of the styled text and use the delete key, styling is NOT retained. Similarly, if you select all the letters of the styled text and delete them, the styling is NOT retained. But if I use the backspace key to delete the selected styled text, the styling IS retained.

This might be a possible solution. There is already a mechanism for recognizing when newlines are inserted (that is actually obsolete now and probably should be deprecated). We might be able to use the same mechanism to recognize when backspace is used.

The problem is that if the selection extends beyond the styled text, backspace does remove the styling. ARGH!
So, we would need to hold the styling of the selection before the backspace event. That adds even more complexity.

Any thoughts? Are we getting closer to a rule?

**Comment by singerdmx** [comment]
The original styling should be there (code wise)

**Comment by AtlasAutocode** [comment]
Just a quick update.
In reviewing how several word processors handle style changes after deletion events, I have come to the conclusion that this issue is only apparent when the backspace key is used.
This issue is only seen when you are simply backspacing or working with selected text (including spanning text with different styling). Using the delete key or replacing selected text behaves correctly.

Bottom line: I have a fix. I have some other related issues to deal with and need to check for any adverse reactions to the change. I should be able to submit a PR soon.

**Comment by geronimol** [comment]
@AtlasAutocode Thank you for this fix!

**Comment by S2V** [comment]
This still reproduces at the 10.5.5 version

**Comment by AtlasAutocode** [comment]
Can you tell me **_exactly_** what you are doing so I can reproduce your issue.
I need to know what steps and platform. I assume you are using the example app.
Obviously, you are doing something different from what we are testing.

**Comment by AtlasAutocode** [comment]
Thank you.
When you delete the text, if you press delete once more after removing all of the text, then you will clear the formatting.

To be exact:
Text = j (shows with formatting)
Press delete
Text = blank (but formatting intact and enter text will be formatted)
**_Press delete again_**
Text = blank (formatting cleared)

Is this a possibility?

**Comment by S2V** [comment]
Yep, your logic is correct, and the controller's listener shows the same thing(my custom toolbar shows the correct typing attributes after deleting the last character using a single 'delete' press)—but if I enter the text, the formatting automatically changes to the default text style, even after a single delete press.

1. The editor contains single 'J' character
2. press 'delete'
3. editor input is empty, and toolbar shows correct/expected formatting for upcoming typing 
4. press any key(M, for ex.) to enter text
5. the formatting automatically resets to default, and character(M) from the previous step is displayed with default attributes(the toolbar also shows this).

So, how to prevent this behavior? I guess https://github.com/singerdmx/flutter-quill/pull/2125 should has been fix this.

**Comment by AtlasAutocode** [comment]
I think there may be a discrepancy.
On windows there is delete and backspace. #2125 dealt with backspace operation (backwards deletion).
I only occasionally use a mac so assumed when you said delete, you were entering the text and then removing with a backspace operation. 

The key issue may be the cursor position when text is removed.
Enter styled text and with the cursor at the end, remove text backwards => style is maintained ('M' is styled)
However, enter styled text and move the cursor to the START, remove text => style is removed ('M' is plain)

It's also (probable?) that the mac implements its key actions differently. For example, if the delete action was a move cursor left and delete the following character that would explain your effect.

Computers are so persnickety! They do exactly what we say, and we have to cope with every combination.
Just to make things more interesting, we have a related issue 2180 that wants the style to be removed which is the opposite of what you want.

I will try to look into this, but I rarely use a mac.

**Comment by AtlasAutocode** [comment]
OK, I got time on a mac.
Cmd+A selects all text and deletes everything including the style. This is expected.

When I enter formatted text and then use the Delete key, it removes characters backwards to the left. The behavior is exactly as I described previously. Formatting is retained when the last character is removed. Pressing delete/backspace one more time will erase the style.

There appears to be a great deal of confusion about the delete key on macs. On my MacBook Air, there is a single delete key that acts like a backspace key removing characters backwards. However, I have seen images of mac keywords with 2 delete keys and they act differently!

Can you confirm that you are deleting backwards and are not pressing delete one more time?
Can you also run the Flutter Quill editor example program and see if you get the same issue?
You may also want to update to the latest quill editor release.

At the moment I cannot reproduce an error.
